### PR TITLE
feat: Ray Cluster and Operator Deployment

### PR DESCRIPTION
**Pull request by MichaelClifford**
## Description
This PR is a follow up to #573. It Includes the manifests needed to deploy Ray on Openshift, but has removed any dependencies on JupyterHub from the previous PR. 

## Merge criteria:

- [x] The commits are squashed in a cohesive manner and have meaningful messages.
- [x] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [x] The developer has manually tested the changes and verified that the changes work

It can be tested by following the instructions in the README included in the PR.

### CODE REVIEW

**Comment by LaVLaS** [comment]
@MichaelClifford This is going to require support CI tests that will run within our test framework. See the test [README](https://github.com/opendatahub-io/odh-manifests/tree/master/tests)

**Review by anishasthana — COMMENTED** [review]
Under what SIG/WG are we proposing this component? Or is it going to be something like a tier 2 component for the time being?

**Comment by MichaelClifford** [comment]
@anishasthana based on the conversation at the ML Ops SIG meeting earlier this week, this will exist initially as a tier 2 component.

**Comment by MichaelClifford** [comment]
@LaVLaS added to the PR to add a couple of tests. Let me know if it needs anything else :smile:

**Comment by LaVLaS** [comment]
> @LaVLaS added to the PR to add a couple of tests. Let me know if it needs anything else smile

@MichaelClifford We are in the process of re-organizing components into tiers.  I know there have been discussions about incubating Ray.io as a Tier 1 component which means it will need a full testsuite to verify:

1. It deploys successfully
2. Test of functionality to ensure that a successful deployment actually works.

You already have some tests to verify that it was deployed successfully, you'll need to add some tests to verify functionality so that any reviewer can determine if any future changes actually work.

**Comment by MichaelClifford** [comment]
@LaVLaS I've added an additional test that checks for functionality. It connects to the example ray cluster, runs a small job, then confirms the job ran on both the head and worker node. 

Thanks for your guidance on this :smiley: Let me know if we need more tests here.

File: ray/cluster/base/kustomization.yaml [hunk 1]
```diff
@@ -0,0 +1,5 @@
+---
+kind: Kustomization
+apiVersion: kustomize.config.k8s.io/v1beta1
+resources:
+  - ray-cluster.yaml
```
Comment (anishasthana) [inline ray/cluster/base/kustomization.yaml:5]: Missing new line

File: ray/cluster/base/ray-cluster.yaml [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+              # The maximum memory that this pod is allowed to use. The
+              # limit will be detected by ray and split to use 10% for
+              # redis, 30% for the shared memory object store, and the
+              # rest for application memory. If this limit is not set and
+              # the object store size is not set manually, ray will
+              # allocate a very large object store in each pod that may
+              # cause problems for other pods.
+              memory: 2G
+  - name: worker-nodes
+    # we can parameterize this when we fix the JH launcher json/jinja bug
```
Comment (anishasthana) [inline ray/cluster/base/ray-cluster.yaml:71]: Given ODH focus on NBC is this still valid?

File: tests/basictests/ray.sh [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+  os::cmd::try_until_text "oc get pods -l app=ray-simple-test -o jsonpath='{$.items[*].status.containerStatuses[0].restartCount}'" "0" $odhdefaulttimeout $odhdefaultinterval
+  pod_name=($(oc get pods -l app=ray-simple-test -o jsonpath='{$.items[*].metadata.name}'))
+  os::cmd::try_until_text "oc logs ${pod_name} | grep 'Simple tests passed'" "Simple tests passed" $odhdefaulttimeout $odhdefaultinterval
+}
+
+check_operator
+check_cluster
+check_functionality
+
+os::test::junit::declare_suite_end
```
Comment (anishasthana) [inline tests/basictests/ray.sh:48]: Missing new line

File: tests/resources/ray/ray-simple-test.yaml [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+    matchLabels:
+      app: ray-simple-test
+  template:
+    metadata:
+      labels:
+        app: ray-simple-test
+    spec:
+      containers:
+      - name: ray-odh-tests
+        image: quay.io/michaelclifford/ray-odh-tests:latest
```
Comment (anishasthana) [inline tests/resources/ray/ray-simple-test.yaml:20]: Ideally we'd have this image in the ODH org. Where's the dockerfile for this image defined?

File: ray/operator/base/ray-operator-deployment.yaml [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+    metadata:
+      labels:
+        app: ray-operator
+    spec:
+      restartPolicy: Always
+      serviceAccountName: ray-operator
+      containers:
+        - name: ray-operator
+          imagePullPolicy: Always
+          image: 'quay.io/thoth-station/ray-operator:v0.1.3'
```
Comment (anishasthana) [inline ray/operator/base/ray-operator-deployment.yaml:22]: Ideally we'd have this image in the ODH org. Where's the dockerfile for this image defined?

File: ray/cluster/base/ray-cluster.yaml [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+      spec:
+        restartPolicy: Never
+        volumes:
+        - name: dshm
+          emptyDir:
+            medium: Memory
+        containers:
+        - name: ray-node
+          imagePullPolicy: Always
+          image: quay.io/thoth-station/ray-ml-worker:v0.2.1
```
Comment (anishasthana) [inline ray/cluster/base/ray-cluster.yaml:89]: Ideally we'd have this image in the ODH org. Where's the dockerfile for this image defined?

File: ray/cluster/base/ray-cluster.yaml [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
```
Comment (anishasthana) [inline ray/cluster/base/ray-cluster.yaml:37]: Ideally we'd have this image in the ODH org. Where's the dockerfile for this image defined?

**Review by anishasthana — COMMENTED** [review]
Couple of nits + things to follow up on eventually. I'll try testing this PR by the end of the week

**Review by anishasthana — CHANGES_REQUESTED** [review]
Actually.... we need metrics and monitoring defined (if available).
I'd take a look at the ModelMesh component for a good example.

https://github.com/opendatahub-io/odh-manifests/tree/master/model-mesh/prometheus
https://github.com/opendatahub-io/odh-manifests/blob/master/tests/basictests/modelmesh.sh#L96

File: tests/resources/ray/ray-simple-test.yaml (hunk 4 above)
Comment (MichaelClifford) [reply tests/resources/ray/ray-simple-test.yaml:20]: Dockerfile is currently here: https://github.com/redhat-et/ray-on-the-edge-demo/tree/main/tests 

Should these files be included in this repo under something like `ray/tests/`? 

What it the process to get images into the ODH org?

**Comment by LaVLaS** [comment]
@MichaelClifford Adding the comment based on the offline conversation with @opendatahub-io/sig-mlops that we will merge this here but include it in the migration to the opendatahub-io-contrib (Tier 2) org since the @opendatahub-io/sig-mlops does not have ray.io on the immediate roadmap for MLOps integration

File: ray/cluster/prometheus/monitor.yaml [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# Prometheus Monitor Service (Metrics)
+apiVersion: monitoring.coreos.com/v1
+kind: ServiceMonitor
+metadata:
+  labels:
+    k8s-app: ray-monitor
+  name: ray-monitor
+  namespace: open-data-hub
```
Comment (anishasthana) [inline ray/cluster/prometheus/monitor.yaml:21]: I don't think this namespace is valid or needed.

File: ray/cluster/prometheus/monitor.yaml [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+spec:
+  endpoints:
+    - interval: 30s
+      port: metrics
+      scheme: http
+  selector:
+    matchLabels:
+      app: ray-monitor
```
Comment (anishasthana) [inline ray/cluster/prometheus/monitor.yaml:29]: Missing newline :-)

**Review by anishasthana — CHANGES_REQUESTED** [review]

File: ray/cluster/prometheus/monitor.yaml (hunk 9 above)
Comment (HumairAK) [inline ray/cluster/prometheus/monitor.yaml:29]: remove newline

File: ray/cluster/base/kustomization.yaml [hunk 10]
```diff
@@ -0,0 +1,7 @@
...
+  - ../prometheus
+
```
Comment (HumairAK) [inline ray/cluster/base/kustomization.yaml:7]: remove newline

File: ray/OWNERS [hunk 11]
```diff
@@ -0,0 +1,9 @@
+approvers:
+- michaelclifford
+- rimolive
+- HumairAK
+
+reviewers:
+- HumairAK
+- michaelclifford
+- rimolive
```
Comment (HumairAK) [inline ray/OWNERS:9]: there's an additional space before `rimolive`

File: ray/cluster/prometheus/monitor.yaml (hunk 9 above)
Comment (HumairAK) [reply ray/cluster/prometheus/monitor.yaml:29]: sorry -- it's the opposite, newline missing

File: ray/cluster/base/kustomization.yaml (hunk 10 above)
Comment (HumairAK) [reply ray/cluster/base/kustomization.yaml:7]: sorry -- it's the opposite, newline missing

File: ray/README.md [hunk 12]
```diff
@@ -0,0 +1,70 @@
...
+
+_WIP Docs: _
+
+Integration of [Ray](https://docs.ray.io/en/latest/index.html) with Open Data Hub on OpenShift. The ray operator and other components are based on https://docs.ray.io/en/releases-1.13.0/cluster/kubernetes.html
+
+## Components  of the Ray deployment
+
+1. [Ray operator](./operator/ray-operator-deployment.yaml): The operator will process RayCluster resources and schedule ray head and worker pods based on requirements.
+2. [Ray CR](./operator/ray-custom-resources.yaml):  RayCluster Custom Resource (CR) describes the desired state of ray cluster.
+3. [Ray Cluster](./cluster/ray-cluster.yaml): Defines an instance of an example Ray Cluster
```
Comment (HumairAK) [inline ray/README.md:11]: relative links to .yaml files won't work, you'll have to link the resulting url path

File: ray/README.md [hunk 13]
```diff
@@ -0,0 +1,70 @@
+# Deploying Ray with Open Data Hub
+
+_WIP Docs: _
```
Comment (HumairAK) [inline ray/README.md:3]: is this supposed to be here? seems unnecessary

File: ray/README.md [hunk 14]
```diff
@@ -0,0 +1,70 @@
...
+### Install Ray
+
+We will use [Kustomize](https://kustomize.io/) to deploy everything we need to use Ray with Open Data Hub. 
+
+#### Install the operator and custom resource 
+
+First use the `oc kustomize` command to generate a yaml containing all the requirements for the operator and the "raycluster" custom resource, then `oc apply` that yaml to deploy the operator to your cluster. 
+
+```bash
+$ oc kustomize deploy/odh-ray-nbc/operator > operator_deployment.yaml
```
Comment (HumairAK) [inline ray/README.md:31]: `deploy/odh-ray-nbc/operator` 
Is this an outdated path? the current path to the operator seems to be `ray/operator/base` ?

File: ray/README.md [hunk 15]
```diff
@@ -0,0 +1,70 @@
...
+```
+```bash
+$ oc apply -f operator_deployment.yaml
```
Comment (HumairAK) [inline ray/README.md:34]: as this is a namespace scoped deployment I'd specify `-n <your-namespace>`

File: ray/README.md [hunk 16]
```diff
@@ -0,0 +1,70 @@
...
+NAME                               READY   STATUS    RESTARTS        AGE
+ray-operator-867bc855b7-2tzxs      1/1     Running   0               4d19h
+
+```
+
+#### Create a ray cluster 
+
+
+```bash
+$ oc kustomize deploy/odh-ray-nbc/cluster > cluster_deployment.yaml
```
Comment (HumairAK) [inline ray/README.md:50]: same as before, the path `deploy/odh-ray-nbc/cluster` seems like a typo, looks like the cluster is at `ray/cluster/base`

File: ray/README.md (hunk 12 above)
Comment (MichaelClifford) [reply ray/README.md:11]: I think you can. I just need to update the links, these are based on an older file structure.

**Review by LaVLaS — APPROVED** [review]

**Pull request merged**